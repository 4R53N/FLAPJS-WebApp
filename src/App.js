import React from 'react';
import { hot } from 'react-hot-loader';

import HomePage from 'pages/intro/HomePage.js';
import WorkPage from 'pages/content/WorkPage.js';
import Page404 from 'pages/Page404.js';

const PAGES = {
  '/': HomePage,
  '/content': WorkPage
};

class App extends React.Component
{
  render()
  {
    const PageHandler = PAGES[this.props.pathname] || Page404;
    return <PageHandler><PageHandler/>;
  }
}

//For hotloading this class
export default hot(module)(App);
