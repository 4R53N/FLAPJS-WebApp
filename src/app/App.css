#root {
  height: 100%;
}

.app-container {
  display: flex;
  flex-direction: column;
  height: 100%;
}

.toolbar-container{
  z-index: 1000;
}

.workspace-container {
  flex-grow: 1;

  height: calc(100% - 4em);
}

#workspace-content {
  display: block;
  width: 100%;
  height: 100%;
}

.workspace-main {
  position: fixed;
  width: 100%;
  height: inherit;

  transition: right 0.3s ease;

  right: 0;
}

.workspace-viewport {
  z-index: 500;
  position: fixed;
  left: 0;
  height: inherit;

  transition: width 0.3s ease;

  width: 100%;
}

.workspace-drawer {
  z-index: 999;
  position: fixed;
  width: 300px;/* Drawer width */
  height: inherit;

  transition: right 0.3s ease;

  right: -300px;/* -(Drawer width) */
}

.workspace-main.open {
  right: 150px;/* (Drawer width) / 2 */
}
.workspace-viewport.open {
  width: calc(100% - 300px);/* width - (Drawer width) */
}
.workspace-drawer.open {
  right: 0px;
}

@media screen and (max-width: 400px) {
  .workspace-main {
    right: 0px; /* Overwrite changes */

    transition: bottom 0.3s ease;

    bottom: 0px;
  }
  .workspace-viewport {
    transition: height 0.3s ease;

    height: inherit;
  }
  .workspace-drawer {
    right: -300px; /* Overwrite changes */

    width: 100%;
    height: 300px;/* Drawer height */

    /* BUGFIX: when expands, it will not translate the right property*/
    transition:
      right 0.3s step-end,
      bottom 0.3s ease;

    bottom: -300px;/* -(Drawer width) */
  }

  .workspace-main.open {
    right: 0px; /* Overwrite changes */

    bottom: 150px;/* (Drawer width) / 2 */
  }
  .workspace-viewport.open {
    width: 100%; /* Overwrite changes */

    height: calc(100% - 300px - 4em);/* Height - Drawer height - toolbar*/
  }
  .workspace-drawer.open {
    right: 0px; /* Overwrite changes */

    transition: bottom 0.3s ease;

    bottom: 0px;
  }
}

@media screen and (max-height: 300px) {
  /* TODO: it should be auto fullscreen mode */
}
