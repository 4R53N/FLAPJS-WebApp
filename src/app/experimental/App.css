#root {
  width: 100%;
  height: 100%;

  --color-graph-node: #FEE781;
  --color-graph-text: #000000;
  --color-graph-select: #808080;

  --color-accent: #FFCD00;

  --color-primary: #07889B;
  --color-primary-text: #F7F5E6;
  --color-primary-lite: #63ADB8;
  --color-primary-dark: #077787;

  --color-surface: #AAAAAA;
  --color-surface-text: #000000;
  --color-surface-active: #FFFFFF;
  --color-surface-lite: #CCCCCC;
  --color-surface-dark: #777777;

  --color-background: #FFFFFF;
  --color-background-lite: #CCCCCC;
  --color-background-active: #777777;

  --color-error: #F54545;/*#B05454*/
  --color-surface-error: #C77575;
  --color-surface-error-dark: #B54A4A;

  --color-success: #00FF00;
  --color-surface-success: #A1C797;
  --color-surface-success-dark: #468C34;

  --color-warning: #E6CB53;
  --color-surface-warning: #D6A922;
  --color-surface-warning-dark: #AD8715;
}

/*For hiding scrollbars*/
::-webkit-scrollbar {
  display: none;
}
/*
* {
  -ms-overflow-style: none;
}
*/

:local(.app_container) {
  position: relative;
  width: 100vw;
  height: 100vh;
  display: flex;
  flex-direction: column;
  /* TODO: For production, add this back in. */
  overflow: hidden;
}

:local(.app_bar).open ~ :local(.app_content) {
  visibility: hidden;
  transition: visibility 0s linear 0.3s;
}
