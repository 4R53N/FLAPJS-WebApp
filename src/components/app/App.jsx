import React from 'react';
import PropTypes from 'prop-types';
import Style from './App.module.css';
import './Theme.css';

import { LocalizationProvider } from '@flapjs/util/localization/LocalizationContext.jsx';
import LocaleString from '@flapjs/util/localization/LocaleString.jsx';
import WorkspaceLayout from '../workspace/WorkspaceLayout.jsx';
import DrawerLayout from '../drawer/DrawerLayout.jsx';
import SideBarLayout from '../sidebar/SideBarLayout.jsx';
import { WrenchIcon, RunningManIcon, BarChartIcon } from '../icons/Icons.js';
import FlexibleOrientationLayout from './FlexibleOrientationLayout.jsx';

import Pane from '../panel/pane/Pane.jsx';
import ColorPane from '../panel/colorPane/ColorPane.jsx';

/**
 * A React component that can do anything you want. :D
 */
class App extends React.Component
{
    constructor(props)
    {
        super(props);

        this.state = {
            open: true
        };
        
        this.drawer = React.createRef();
    }

    /** @override */
    render()
    {
        const props = this.props;
        const state = this.state;

        return (
            <div className={Style.container + (props.className || '')}>
                <LocalizationProvider localeCode="en">
                    <nav className={Style.appbar}>
                        <h2>Flap.js</h2>
                        <p><LocaleString entity="hi"/></p>
                        <button onClick={() => this.setState(prev => { return { open: !prev.open }; })}>
                            OPEN
                        </button>
                    </nav>
                    <FlexibleOrientationLayout>
                        {
                            orientation =>
                            {
                                const side = orientation === 'row' ? 'right' : 'bottom';
                                const direction = orientation === 'row' ? 'horizontal' : 'vertical';
                                
                                return (
                                    <WorkspaceLayout
                                        renderBackground={() =>
                                            <div style={{background: 'dodgerblue', width: '100%', height: '100%'}}>
                                            </div>}
                                        renderForeground={() =>
                                            <SideBarLayout
                                                side={side}
                                                renderSideBar = {() =>
                                                    <div className={Style.sidetab + ' ' + direction}>
                                                        <button className={Style.tab} onClick={() => this.setState({open: true})}>
                                                            <RunningManIcon className="icon" color="#FFFFFF"/>
                                                        </button>
                                                        <button className={Style.tab} onClick={() => this.setState({open: true})}>
                                                            <BarChartIcon className="icon" color="#FFFFFF"/>
                                                        </button>
                                                        <button className={Style.tab} onClick={() => this.setState({open: true})}>
                                                            <WrenchIcon className="icon" color="#FFFFFF"/>
                                                        </button>
                                                        <div className={Style.divider}></div>
                                                        <button className={Style.tab} onClick={() => this.setState({open: true})}>
                                                            <BarChartIcon className="icon" color="#FFFFFF"/>
                                                        </button>
                                                    </div>
                                                }>
                                                <DrawerLayout
                                                    ref={this.drawer}
                                                    side={side}
                                                    open={state.open}
                                                    renderDrawer = {() =>
                                                        <div style={{color: 'white', width: '100%', height: '100%'}}>
                                                            <ColorPane source={() => document.querySelector('#root')}>
                                                            </ColorPane>
                                                            <Pane title="About" open={true}>
                                                                <p>I am content :D</p>
                                                                <p>
                                                                    Aenean libero orci, mattis ut ullamcorper eu, scelerisque et augue. Nam egestas vestibulum interdum. Integer ligula nunc, efficitur vitae velit sed, tristique cursus ligula. Aliquam est est, consectetur quis turpis a, luctus ultrices odio. Morbi a feugiat mi. Fusce orci dolor, maximus quis eros non, placerat accumsan tellus. Nam a lectus vitae nisl scelerisque ultrices in vitae erat.
                                                                </p>
                                                                <p>
                                                                    Sed auctor, augue non sodales euismod, metus nulla congue ex, sit amet rhoncus metus turpis ut elit. Sed vitae rutrum augue. Etiam dictum ornare tortor sed tristique. Cras quis velit nisl. Nulla rhoncus lacus sit amet interdum consectetur. Pellentesque aliquam consequat vehicula. Sed eu dignissim eros. Donec rhoncus consectetur ante, sit amet pharetra urna tempor sed. In scelerisque ligula mi. Nam gravida maximus ex, consequat ultricies nisi molestie non. Maecenas id semper eros.
                                                                </p>
                                                            </Pane>
                                                            <Pane title="Hello World" open={true}>
                                                                <p>I am content :D</p>
                                                                <p>
                                                                    Aenean libero orci, mattis ut ullamcorper eu, scelerisque et augue. Nam egestas vestibulum interdum. Integer ligula nunc, efficitur vitae velit sed, tristique cursus ligula. Aliquam est est, consectetur quis turpis a, luctus ultrices odio. Morbi a feugiat mi. Fusce orci dolor, maximus quis eros non, placerat accumsan tellus. Nam a lectus vitae nisl scelerisque ultrices in vitae erat.
                                                                </p>
                                                                <p>
                                                                    Sed auctor, augue non sodales euismod, metus nulla congue ex, sit amet rhoncus metus turpis ut elit. Sed vitae rutrum augue. Etiam dictum ornare tortor sed tristique. Cras quis velit nisl. Nulla rhoncus lacus sit amet interdum consectetur. Pellentesque aliquam consequat vehicula. Sed eu dignissim eros. Donec rhoncus consectetur ante, sit amet pharetra urna tempor sed. In scelerisque ligula mi. Nam gravida maximus ex, consequat ultricies nisi molestie non. Maecenas id semper eros.
                                                                </p>
                                                            </Pane>
                                                            <Pane title="Nothing To See Here" open={true}>
                                                                <p>I am content :D</p>
                                                                <p>
                                                                    Aenean libero orci, mattis ut ullamcorper eu, scelerisque et augue. Nam egestas vestibulum interdum. Integer ligula nunc, efficitur vitae velit sed, tristique cursus ligula. Aliquam est est, consectetur quis turpis a, luctus ultrices odio. Morbi a feugiat mi. Fusce orci dolor, maximus quis eros non, placerat accumsan tellus. Nam a lectus vitae nisl scelerisque ultrices in vitae erat.
                                                                </p>
                                                                <p>
                                                                    Sed auctor, augue non sodales euismod, metus nulla congue ex, sit amet rhoncus metus turpis ut elit. Sed vitae rutrum augue. Etiam dictum ornare tortor sed tristique. Cras quis velit nisl. Nulla rhoncus lacus sit amet interdum consectetur. Pellentesque aliquam consequat vehicula. Sed eu dignissim eros. Donec rhoncus consectetur ante, sit amet pharetra urna tempor sed. In scelerisque ligula mi. Nam gravida maximus ex, consequat ultricies nisi molestie non. Maecenas id semper eros.
                                                                </p>
                                                            </Pane>
                                                        </div>
                                                    }>
                                                    <div style={{color: 'white', width: '100%', height: '100%'}}>
                                                        Lorem ipsum dolor sit amet, consectetur adipiscing elit. Maecenas in sem magna. Aliquam ultrices convallis cursus. In hac habitasse platea dictumst. Quisque eu tellus magna. Curabitur aliquam luctus ex. Maecenas purus arcu, tincidunt non sapien quis, finibus vestibulum ante. Curabitur vulputate et ligula in congue. Phasellus ac imperdiet libero, sagittis interdum elit. Suspendisse potenti.
                                                    </div>
                                                </DrawerLayout>
                                            </SideBarLayout>
                                        }>
                                    </WorkspaceLayout>
                                );
                            }
                        }
                    </FlexibleOrientationLayout>
                </LocalizationProvider>
            </div>
        );
    }
}

App.propTypes = {
    className: PropTypes.string,
};
App.defaultProps = {
};

export default App;
