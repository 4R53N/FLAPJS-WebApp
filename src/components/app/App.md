# App
This is the entrypoint component of the entire app. The only thing higher than this is the literal "root" component with the "root" id (which is managed by `index.js`). Everything is basically connected to here.

## Motivation
You need some central place to organize app-wide communications. And cause I love all of you. <3

## Work Log
- Added layouts to the app, instead of the old throw-anything-in-that-works code blocks.
- Created App component.