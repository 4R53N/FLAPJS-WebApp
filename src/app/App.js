import React from 'react';
import { hot } from 'react-hot-loader';

import './App.css';

import Toolbar from 'toolbar/Toolbar.js';
import Workspace from 'workspace/Workspace.js';
import Drawer from 'workspace/Drawer.js';
import Viewport from 'workspace/Viewport.js';

class App extends React.Component
{
  constructor(props)
  {
    super(props)

    this.state = {
      isOpen: false
    };
  }

  openDrawer()
  {
    this.setState((prev, props) => {
      return { isOpen: true };
    });
  }

  closeDrawer()
  {
    this.setState((prev, props) => {
      return { isOpen: false };
    });
  }

  isDrawerOpen()
  {
    return this.state.isOpen;
  }

  render()
  {
    return (
      <div className="app-container">
        <Toolbar app={this}/>

        <div className="workspace-container">
          <div className={"workspace-main" + (this.state.isOpen ? " open" : "")}>
            <Workspace />
          </div>
          <div className={"workspace-viewport" + (this.state.isOpen ? " open" : "")}>
            <Viewport />
          </div>
          <div className={"workspace-drawer" + (this.state.isOpen ? " open" : "")}>
            <Drawer />
          </div>
        </div>
      </div>
    );
  }
}

//For hotloading this class
export default hot(module)(App);
