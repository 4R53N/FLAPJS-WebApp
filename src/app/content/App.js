import React from 'react';
import { hot } from 'react-hot-loader';
import './App.css';

import NodalGraph from 'graph/NodalGraph.js';

import GraphController from 'controller/GraphController.js';
import InputController from 'controller/InputController.js';
import MachineController from 'controller/MachineController.js';

import AutoSaver from 'util/AutoSaver.js';
import HotKeys from './HotKeys.js';

import Toolbar from './toolbar/Toolbar.js';
import Workspace from './workspace/Workspace.js';
import Drawer from './drawer/Drawer.js';
import Viewport from './viewport/Viewport.js';
import NotificationSystem from 'notification/NotificationSystem.js';
import Tutorial from 'tutorial/Tutorial.js';

import FSABuilder from 'builder/FSABuilder.js';
import GraphUploader from 'graph/GraphUploader.js';
import TestingManager from 'testing/TestingManager.js';
import EventManager from './EventManager.js';

class App extends React.Component
{
  constructor(props)
  {
    super(props);

    //Create references
    this.container = null;
    this.workspace = null;
    this.viewport = null;
    this.drawer = null;
    this.notification = null;
    this.toolbar = null;

    this._graph = new NodalGraph();
    this._machineBuilder = new FSABuilder(this._graph);

    //Must be initialized (will be called in Workspace.componentDidMount)
    this.graphController = new GraphController(this._graph);
    this.inputController = new InputController(this._graph);
    this.machineController = new MachineController(this._machineBuilder);

    this.testingManager = new TestingManager();
    this.eventManager = new EventManager();

    this.hotKeys = new HotKeys();
    this.tutorial = new Tutorial();

    this.state = {
      isOpen: true,
      isDangerous: false,
      isFullscreen: false,
      isWaitingForFile: false
    };

    this.onDragOver = this.onDragOver.bind(this);
    this.onDragEnter = this.onDragEnter.bind(this);
    this.onDragLeave = this.onDragLeave.bind(this);
    this.onFileDrop = this.onFileDrop.bind(this);
  }

  componentDidMount()
  {
    //Does the browser support autosaving?
    if(AutoSaver.doesSupportLocalStorage())
    {
      AutoSaver.loadAutoSave(this._graph);

      //Start auto-saving
      AutoSaver.initAutoSave(this._graph);
    }

    //Initialize the controller to graph components
    this.inputController.initialize(this);
    this.graphController.initialize(this);
    this.machineController.initialize(this);

    this._machineBuilder.initialize(this);
    this.testingManager.initialize(this);
    this.eventManager.initialize(this);
    this.hotKeys.initialize(this);

    //Notify on create in delete mode
    const tryCreateWhileTrash = () => {
      if (this.inputController.getPointer().trashMode)
      {
        this.notification.addWarningMessage(I18N.toString("message.warning.cannotmodify"),
          "tryCreateWhileTrash", true);
      }
    };
    this.graphController.on("tryCreateWhileTrash", tryCreateWhileTrash);

    //Upload drop zone
    const workspaceDOM = this.workspace.ref;
    workspaceDOM.addEventListener("drop", this.onFileDrop);
    workspaceDOM.addEventListener("dragover", this.onDragOver);
    workspaceDOM.addEventListener("dragenter", this.onDragEnter);
    workspaceDOM.addEventListener("dragleave", this.onDragLeave);

    //Begin tutorial
    this.tutorial.start(this);
  }

  componentWillUnmount()
  {
    this.hotKeys.destroy();
    this.eventManager.destroy();

    this.machineController.destroy();
    this.graphController.destroy();
    this.inputController.destroy();

    this._machineBuilder.destroy();

    //Upload drop zone
    const workspaceDOM = this.workspace.ref;
    workspaceDOM.removeEventListener("drop", this.onFileDrop);
    workspaceDOM.removeEventListener("dragover", this.onDragOver);
    workspaceDOM.removeEventListener("dragenter", this.onDragEnter);
    workspaceDOM.removeEventListener("dragleave", this.onDragLeave);
  }

  //Called to prevent default file open
  onDragOver(ev)
  {
    ev.preventDefault();
  }

  onDragEnter(ev)
  {
    ev.preventDefault();

    //Change state
    this.setState({ isWaitingForFile: true });
    this.viewport.setState((prev, props) => {
      return {
        prevMode: prev.mode,
        mode: Viewport.DANGEROUS
      };
    });
  }

  onDragLeave(ev)
  {
    ev.preventDefault();

    //Revert state
    this.setState({ isWaitingForFile: false });
    this.viewport.setState((prev, props) => {
      return {
        prevMode: prev.mode,
        mode: prev.prevMode
      };
    });
  }

  onFileDrop(ev)
  {
    //Prevent file from being opened
    ev.preventDefault();

    //Revert state
    this.setState({ isWaitingForFile: false });
    this.viewport.setState((prev, props) => {
      return {
        prevMode: prev.mode,
        mode: prev.prevMode
      };
    });

    if (ev.dataTransfer.items)
    {
      const length = ev.dataTransfer.items.length;
      for(let i = 0; i < length; ++i)
      {
        let file = ev.dataTransfer.items[i];
        if (file.kind === 'file')
        {
          const data = file.getAsFile();
          //TODO: should show error message if invalid
          GraphUploader.uploadFileToGraph(data, this._graph, null, null);
        }
      }
    }
    else
    {
      const length = ev.dataTransfer.files.length;
      for(let i = 0; i < length; ++i)
      {
        const data = ev.dataTransfer.files[i];
        //TODO: should show error message if invalid
        GraphUploader.uploadFileToGraph(data, this._graph, null, null);
      }
    }

    if (ev.dataTransfer.items)
    {
      ev.dataTransfer.items.clear();
    }
    else
    {
      ev.dataTransfer.clearData();
    }
  }

  openDrawer(full=null)
  {
    if (typeof full === 'boolean')
    {
      this.setState((prev, props) => {
        return { isOpen: true, isFullscreen: full };
      });
    }
    else
    {
      this.setState((prev, props) => {
        return { isOpen: true };
      });
    }
  }

  closeDrawer()
  {
    this.setState((prev, props) => {
      return { isOpen: false };
    });
  }

  isDrawerOpen()
  {
    return this.state.isOpen;
  }

  shouldHideContent()
  {
    return this.state.isFullscreen && this.state.isOpen;
  }

  render()
  {
    const inputController = this.inputController;
    const graphController = this.graphController;
    const machineController = this.machineController;

    const graph = this._graph;
    const machineBuilder = this._machineBuilder;
    const tester = this.testingManager;
    const screen = this.workspace ? this.workspace.ref : null;

    inputController.update();

    return <div className="app-container" ref={ref=>this.container=ref}>
      <Toolbar ref={ref=>this.toolbar=ref}
        graphController={graphController}
        machineController={machineController}
        eventManager={this.eventManager}
        drawer={this.drawer}
        notification={this.notification}/>

      <NotificationSystem ref={ref=>this.notification=ref}
        graphController={graphController}
        machineController={machineController}/>

      <div className="workspace-container">
        <div className={"workspace-main" +
          (this.state.isOpen ? " open" : "")}
          style={{
            visibility: this.shouldHideContent() ? "hidden" : "visible",
            opacity: this.state.isWaitingForFile ? "0.1" : "1"
          }}>

          <Workspace ref={ref=>this.workspace=ref}
            graphController={graphController}
            inputController={inputController}
            machineController={machineController}
            tester={tester}/>
        </div>

        <div className={"workspace-viewport" +
          (this.state.isOpen ? " open" : "")}
          style={{
            visibility: this.shouldHideContent() ? "hidden" : "visible"
          }}>

          <Viewport ref={ref=>this.viewport=ref}
            graphController={graphController}
            inputController={inputController}
            machineController={machineController}
            tester={tester}
            screen={screen}/>
        </div>

        <div className={"workspace-drawer" +
          (this.state.isOpen ? " open" : "") +
          (this.state.isFullscreen ? " fullscreen" : "")}>

          <Drawer ref={ref=>this.drawer=ref}
            app={this}
            graphController={graphController}
            machineController={machineController}
            toolbar={this.toolbar} />
        </div>
      </div>
    </div>;
  }
}

//For hotloading this class
export default hot(module)(App);
