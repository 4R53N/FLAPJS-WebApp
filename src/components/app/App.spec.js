/* eslint-disable import/first */
import React from 'react';
import { shallow } from 'enzyme';

jest.mock('@flapjs/util/MathHelper.js', () => ({
    __esModule: true,
    uuid: jest.fn(),
}));

import { uuid } from '@flapjs/util/MathHelper.js';
import App from './App.jsx';

let wrapper;
let nextUUID = 0;

beforeEach(() =>
{
    uuid.mockImplementation(() => nextUUID++);
    wrapper = shallow(<App/>);
});

describe('<App/> rendering', () =>
{
    test('renders correctly', () =>
    {
        expect(wrapper).toMatchSnapshot();
    });
});

describe('<App/> interaction', () =>
{
    /** I promise I will test you soon... */
});

describe('<App/> lifecycle', () =>
{
    /** ...later. */
});
