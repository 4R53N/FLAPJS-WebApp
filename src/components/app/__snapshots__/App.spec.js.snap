// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App/> rendering renders correctly 1`] = `
<div
  className="container"
>
  <SessionProvider
    onDidMount={[Function]}
    onLoad={[Function]}
    onUnload={[Function]}
    onWillUnmount={[Function]}
    reducer={[Function]}
    renderChildren={[Function]}
    state={Object {}}
  >
    <AppServiceProviders
      appProps={
        Object {
          "changeModule": [Function],
          "module": null,
          "renderModule": [Function],
        }
      }
    >
      <DeprecatedServiceProviders
        app={
          App {
            "context": Object {},
            "moduleSessionHandler": ModuleSessionHandler {
              "changeModuleCallback": [Function],
              "module": null,
              "onDidMount": [Function],
              "onLoad": [Function],
              "onUnload": [Function],
              "onWillUnmount": [Function],
              "reducer": [Function],
            },
            "props": Object {
              "changeModule": [Function],
              "module": null,
              "renderModule": [Function],
            },
            "refs": Object {},
            "setState": [Function],
            "state": null,
            "updater": Updater {
              "_callbacks": Array [],
              "_renderer": ReactShallowRenderer {
                "_context": Object {},
                "_didScheduleRenderPhaseUpdate": false,
                "_dispatcher": Object {
                  "readContext": [Function],
                  "useCallback": [Function],
                  "useContext": [Function],
                  "useDebugValue": [Function],
                  "useDeferredValue": [Function],
                  "useEffect": [Function],
                  "useImperativeHandle": [Function],
                  "useLayoutEffect": [Function],
                  "useMemo": [Function],
                  "useReducer": [Function],
                  "useRef": [Function],
                  "useResponder": [Function],
                  "useState": [Function],
                  "useTransition": [Function],
                },
                "_element": <App
                  changeModule={[Function]}
                  module={null}
                  renderModule={[Function]}
                />,
                "_firstWorkInProgressHook": null,
                "_forcedUpdate": false,
                "_instance": [Circular],
                "_isReRender": false,
                "_newState": null,
                "_numberOfReRenders": 0,
                "_renderPhaseUpdates": null,
                "_rendered": <div
                  className="container"
                >
                  <SessionProvider
                    onDidMount={[Function]}
                    onLoad={[Function]}
                    onUnload={[Function]}
                    onWillUnmount={[Function]}
                    reducer={[Function]}
                    renderChildren={[Function]}
                    state={Object {}}
                  >
                    <AppServiceProviders
                      appProps={
                        Object {
                          "changeModule": [Function],
                          "module": null,
                          "renderModule": [Function],
                        }
                      }
                    >
                      <DeprecatedServiceProviders
                        app={[Circular]}
                      >
                        <AppBar>
                          <SessionStateConsumer>
                            [Function]
                          </SessionStateConsumer>
                        </AppBar>
                        <AppWorkspace
                          renderDrawer={[Function]}
                          renderPlayground={[Function]}
                          renderViewport={[Function]}
                        />
                      </DeprecatedServiceProviders>
                    </AppServiceProviders>
                  </SessionProvider>
                </div>,
                "_rendering": false,
                "_updater": [Circular],
                "_workInProgressHook": null,
              },
            },
          }
        }
      >
        <AppBar>
          <SessionStateConsumer>
            <Component />
          </SessionStateConsumer>
        </AppBar>
        <AppWorkspace
          renderDrawer={[Function]}
          renderPlayground={[Function]}
          renderViewport={[Function]}
        />
      </DeprecatedServiceProviders>
    </AppServiceProviders>
  </SessionProvider>
</div>
`;
